ee/cron/chunksHostLoads.js

```diff
@@ -0,0 +1,189 @@
+
+/**
+ * Finds the maximum value in the given array.
+ * @param {number[]} arr
+ */
+function arrayMax(arr) {
+  return arr.reduce((a, b) => Math.max(a, b), 0);
+}
+
+module.exports.run = callbackify(async () => {
+  if (
+    !config.runningInEc2 ||
+    !config.chunksLoadBalancerDimensionName ||
+    !config.chunksTargetGroupDimensionName
+  ) {
+    return;
+  }
+
+  const now = Date.now();
+  const cloudwatch = new AWS.CloudWatch();
+
+  const metrics = await cloudwatch
+    .getMetricData({
+      // Use the last 15 minutes of data.
+      // TODO: make this configurable?
+      StartTime: new Date(now - 1000 * 60 * 15),
+      EndTime: new Date(now),
+      MetricDataQueries: [
+        {
+          Id: PAGE_VIEWS_PER_SECOND,
+          MetricStat: {
+            Metric: {
+              Namespace: 'PrairieLearn',
+              MetricName: 'PageViewsPerSecond',
+              Dimensions: [
+                {
+                  Name: 'ServerGroup',
+                  Value: config.groupName,
+                },
+              ],
+            },
+            Period: 10,
+            Stat: 'Maximum',
+          },
+        },
+        {
+          Id: ACTIVE_WORKERS_PER_SECOND,
+          MetricStat: {
+            Metric: {
+              Namespace: 'PrairieLearn',
+              MetricName: 'CurrentJobs',
+              Dimensions: [
+                {
+                  Name: 'Job Type',
+                  Value: 'python_worker_active',
+                },
+              ],
+            },
+            Period: 10,
+            Stat: 'Maximum',
+          },
+        },
+        {
+          Id: LOAD_BALANCER_REQUESTS_PER_MINUTE,
+          MetricStat: {
+            Metric: {
+              Namespace: 'AWS/ApplicationELB',
+              MetricName: 'RequestCount',
+              Dimensions: [
+                {
+                  Name: 'LoadBalancer',
+                  Value: config.chunksLoadBalancerDimensionName,
+                },
+                {
+                  Name: 'TargetGroup',
+                  Value: config.chunksTargetGroupDimensionName,
+                },
+              ],
+            },
+            Period: 60,
+            Stat: 'Sum',
+          },
+        },
+      ],
+    })
+    .promise();
+
+  const pageViewsPerSecondMetric = metrics.MetricDataResults.find(
+    (m) => m.Id === PAGE_VIEWS_PER_SECOND
+  );
+  const activeWorkersPerSecondMetric = metrics.MetricDataResults.find(
+    (m) => m.Id === ACTIVE_WORKERS_PER_SECOND
+  );
+  const loadBalancerRequestsPerMinuteMetric = metrics.MetricDataResults.find(
+    (m) => m.Id === LOAD_BALANCER_REQUESTS_PER_MINUTE
+  );
+
+  const maxPageViewsPerSecond = Math.ceil(arrayMax(pageViewsPerSecondMetric.Values));
+  const maxActiveWorkersPerSecond = Math.ceil(arrayMax(activeWorkersPerSecondMetric.Values));
+  const maxLoadBalancerRequestsPerMinute = Math.ceil(
+    arrayMax(loadBalancerRequestsPerMinuteMetric.Values)
+  );
```

Comment:
Why do we want `ceil()` here? I don't think we care if these are not integers.

Note that if we use `_.max()` instead of `arrayMax()` then we should add `?? 0` to the end to handle the zero-length array case.

Reply:
Fair point - I can move the `Math.ceil` call around the division expressions that get assigned to the `desiredInstancesBy...` variables?

Reply:
Changed in 2f3c116, does that look better?

Reply:
Yep! This looks good. I would slightly lean towards computing the individual metrics without the ceil (and applying one ceil at the very end) so that we can see when the individual metrics might be close to an integer boundary. E.g., if we aren't scaling above 1 instance, I'd like to know whether the individual metrics are at 0.8 or 0.008 to see whether we need a small or large change to the capacity factors. However, in practice this probably won't be a huge deal so I'm also fine to leave the code as is.

Reply:
I made this change in 250cd68a88d2db5a8b9e3f419b9165b69212f3cc